### DISCUSSION TITLE
Confused about `--preview`

### DISCUSSION THREAD

**Discussion by buhtz** [astral-sh/ruff#10153] [General]
I am confused about the [Preview Mode](https://docs.astral.sh/ruff/preview/#enabling-preview-mode) using `--preview` on shell.

Referencing this documentation I would say it is a flag to activate several extra linter rules (that are not stable currently).

But it also seems that it modifies the output of ruff. No I get output like this

```
tests/test_report_output.py:14:8: F401 [*] `seaborn` imported but unused
   |
12 | import PIL
13 | import PIL.Image
14 | import seaborn
   |        ^^^^^^^ F401
15 | import docx
16 | import docx.document
   |
   = help: Remove unused import: `seaborn`
```

Instead of just

    tests/test_report_output.py:14:8: F401 [*] `seaborn` imported but unused

I prefer the one-line output but also prefer the extra linter rules.

**Comment by MichaReiser** [comment]
I think there's an opportunity to improve our documentation. The beginning mentions that it is a way for us to receive feedback from the community (phrased openly), but we then narrow the list to rules and formatter changes. 

Our intention is to use `--preview` for all kind of changes, including changing the output format. You can get the "old" output format by using `output-format = "concise"` in your configuration or CLI

**Reply by buhtz** [reply]
So am I right to understand that `--preview` has two effects: 1) modifying the output format and 2) enabling additional rules?

**Reply by MichaReiser** [reply]
Yes. Preview impacts the entire behavior of ruff and not just rules.

**Reply by buhtz** [reply]
Your answer is not correct or not clear enough. The same is with the docs.
As you can see [here](https://github.com/astral-sh/ruff/issues/2402#issuecomment-1970669779) `--preview` do not activate the E303 rule. I have to explicit select it. But `--select E303` alone is not allowed and an error message comes up that I have to use it in combination with `--preview`.

See this example file:

```
# foo.py
class Foo:
    def __init__(self):
        pass


    def two(self):
        pass
```

The E303 is not detected:

    $ ruff foo.py --preview --output-format="concise"
    $ 

This is very confusing from a users perspective.

Let me know if I can help you somehow and testing an improved doc or a modified behavior. I have some ideas.

**Reply by MichaReiser** [reply]
> Your answer is not correct or not clear enough. The same is with the docs.

I'm sorry that my explanation wasn't clear. All I'm saying is that preview mode isn't limited to rules or how your code gets formatted. It also impacts other behavior of ruff, e.g. the default value for `--output-format`. You can think of `--preview` as a way to opt-in to use a future version of ruff. You get access to new commands, rules, formatting styles, but also to changed behavior and defaults.

> As you can see https://github.com/astral-sh/ruff/issues/2402#issuecomment-1970669779 --preview do not activate the E303 rule. I have to explicit select it. But --select E303 alone is not allowed and an error message comes up that I have to use it in combination with --preview.

Using `--preview` does not automatically select additional rules. The default rule selector is `["E4", "E7", "E9", "F"]` and it does not include `E303`. See [Using preview rules](https://docs.astral.sh/ruff/preview/#using-rules-that-are-in-preview) for more details on how to enable preview rules

Only using `--select E303` without preview doesn't enable the rule for you because we want the opt-in to preview rules to be explicit. You should know that you're using less-tested rules and not accidentally run into them.


> Let me know if I can help you somehow and testing an improved doc or a modified behavior. I have some ideas.

I would appreciate that a lot. Do you want to PR a proposal with your improvements?

**Reply by hoel-bagard** [reply]
@MichaReiser I'm thinking that having a CLI command example tab next to the `pyproject.toml` and `ruff.toml` tabs might make it less confusing for people trying out ruff through the CLI, what do you think ? Would it be ok if I opened a PR to add it ?

(Maybe incorporating your answer)

**Reply by MichaReiser** [reply]
@hoel-bagard you can try but I'm not sure if that's easily possible without having to change all examples to have a CLI tab too

**Reply by buhtz** [reply]
It might depend on my none-native English but also the doc text is misleading. I read it several times and nI still understand that `--preview` will activate all new/unstable rules.

When you look into the docs about E303 you are also guided to the preview mode docu.

I still do not understand the concept behind all this. Having unstable rules is a great idea. But I need to activate them somehow **without** explicit select each one of them. I understand `--preview` is the switch for this.
If not why not introduce a new "label" named `UNSTABLE` that can be used like `ALL` with `--select`?

    $ ruff --select=UNSTABLE *.py

**Reply by hoel-bagard** [reply]
I've tried it, and it's possible to add a tab to a single example.

**Reply by hoel-bagard** [reply]
@buhtz I'll try to reformulate it and open a PR, is it ok if I ping you in the PR to see if you find it less confusing ?

**Reply by hoel-bagard** [reply]
> If not why not introduce a new "label" named UNSTABLE that can be used like ALL with --select?

Using `ALL` in combination with the `preview` mode will activate all the preview rules.

![image](https://github.com/astral-sh/ruff/assets/34478245/a4625d35-4c36-41f5-b8bc-25b5871cd26e)

**Reply by buhtz** [reply]
Yes, you can ping me.

Can you explain the term "preview mode" please?
Is this just another output format or does this influence the selection of rules?

Is `preview = true` in pyproject.toml something different from `--preview` on shell?

**Reply by hoel-bagard** [reply]
By "preview mode" I meant using `--preview`.

> Is `preview = true` in pyproject.toml something different from `--preview` on shell?

They have the same effect.

**Reply by buhtz** [reply]
OK, then what is the "effect"? Regarding my experiement the effect is only another output format but **not** select any rules. Am I right?

**Reply by hoel-bagard** [reply]
Rather than selecting the rules, it allows you to select preview only rules. So by itself it doesn't do much, it mainly enables more options.

If you happened to have `ALL` rules enabled, then it would have the effect of turning on all the preview rules.

**Reply by buhtz** [reply]
"it allows you to select preview only rules" -> Easy to misunderstand by none-native speakers.

I understand but that behavior is misleading and confusing.
There  is a switch/option to modify the output format. But indirect it kind of activates hidden selection options.

That is IMHO not a good design and IMHO a bug. The "connection" between `--preview` (modify output format) and rule selection should be killed.
